beacon_node/beacon_chain/tests/rewards.rs

```diff
@@ -219,6 +219,75 @@ async fn test_verify_attestation_rewards_base_inactivity_leak() {
     assert_eq!(expected_balances, balances);
 }
 
+#[tokio::test]
+async fn test_verify_attestation_rewards_altair_inactivity_leak() {
+    let spec = ForkName::Altair.make_genesis_spec(E::default_spec());
+    let harness = get_harness(spec.clone());
+
+    let half = VALIDATOR_COUNT / 2;
+    let half_validators: Vec<usize> = (0..half).collect();
+    // target epoch is the epoch where the chain enters inactivity leak
+    let target_epoch = &spec.min_epochs_to_inactivity_penalty + 2;
+
+    // advance until beginning of epoch N + 1 and get balances
+    harness
+        .extend_chain(
+            (E::slots_per_epoch() * (target_epoch + 1)) as usize,
+            BlockStrategy::OnCanonicalHead,
+            AttestationStrategy::SomeValidators(half_validators.clone()),
+        )
+        .await;
+    let initial_balances: Vec<u64> = harness.get_current_state().balances().clone().into();
+
+    // extend slots to beginning of epoch N + 2
+    harness.advance_slot();
+    harness
+        .extend_chain(
+            E::slots_per_epoch() as usize,
+            BlockStrategy::OnCanonicalHead,
+            AttestationStrategy::SomeValidators(half_validators),
+        )
+        .await;
+    let _slot = harness.get_current_slot();
+
+    // compute reward deltas for all validators in epoch N
+    let StandardAttestationRewards {
+        ideal_rewards,
+        total_rewards,
+    } = harness
+        .chain
+        .compute_attestation_rewards(Epoch::new(target_epoch), vec![])
+        .unwrap();
+
+    // assert inactivity penalty for both ideal rewards and individual validators
+    assert!(ideal_rewards.iter().all(|reward| reward.inactivity == 0));
+    assert!(total_rewards[..half]
+        .iter()
+        .all(|reward| reward.inactivity == 0));
+    assert!(total_rewards[half..]
+        .iter()
+        .all(|reward| reward.inactivity < 0));
+
+    // apply attestation rewards to initial balances
+    let expected_balances = apply_attestation_rewards(&initial_balances, total_rewards);
+
+    for initial_balance in &initial_balances {
+        println!("initial {initial_balance}");
+    }
+
+    for expected_balance in &expected_balances {
+        println!("expected {expected_balance}");
+    }
+    // verify expected balances against actual balances
+    let balances: Vec<u64> = harness.get_current_state().balances().clone().into();
+
+    for balance in &balances {
+        println!("actual {balance}");
+    }
+    // Failing test statement
+    assert_eq!(expected_balances, balances);
```

Comment:
@jimmygchen I am getting a failure at this assertion statement. I have println! to show initial_balances, expected_balances, and the current state balances. The initial_balances to expected_balances change seem to make sense when looking at the deltas, but the results varies a lot when comparing expected_balances to the balances. 

I am wondering why this calculation would be off so much?

Reply:
@jimmygchen the extend_chain function in the test_utils.rs file looks like the attestation reward is being applied when an inactivity leak is on going and potentially throwing off this assert_eq!(expected_balances, balances);

Reply:
It looks like adding the `process_inactivity_updates` logic to the [beginning of the calculation](https://github.com/sigp/lighthouse/blob/70084693950d7460ddc31428c64e05d8697bc8a6/beacon_node/beacon_chain/src/attestation_rewards.rs#L125) gets us a bit closer - I'm seeing matching numbers for _some_ validators now, which is a good sign!

Reply:
@jimmygchen I have the initial balances retrieved for slot 56 and apply the compute_attestation_rewards and slot 64. I have println functions running for initial balances, expected balances(apply_attestation_rewards), and actual balances. I have the results printed out for validator in position 0:

```
Validator: 0 | initial value 32003025960 | slot 56
Validator: 0 | expected value 32003025960 | slot 64
Validator: 0 | actual value 32003798995 | slot 64
```

The first item **Validator: 0 | initial value 32003025960 | slot 56** is calculated and printed as:
https://github.com/sigp/lighthouse/blob/9d866073803032a1b4689e7dff65f8aca00becbf/beacon_node/beacon_chain/tests/rewards.rs#L241-L247
Second item **Validator: 0 | expected value 32003025960 | slot 64** is calculated and printed as:
https://github.com/sigp/lighthouse/blob/9d866073803032a1b4689e7dff65f8aca00becbf/beacon_node/beacon_chain/tests/rewards.rs#L279-L284
Third item **Validator: 0 | actual value 32003798995 | slot 64** is calculated and printed as:
https://github.com/sigp/lighthouse/blob/9d866073803032a1b4689e7dff65f8aca00becbf/beacon_node/beacon_chain/tests/rewards.rs#L286-L291

I believe the actual value **Validator: 0 | actual value 32003798995 | slot 64** calculated on L286-291 in the example above should not have increase if we are in inactivity leak. The slots per epoch is 8, so I believe this would be differences between epoch 7 and 8.

Reply:
ah right, looks like those are the _proposal rewards_! That explains why only _some_ of the balances don't match after adding the `process_inacitivity_updates`.

The reason why the phase0 tests work is because proposer rewards were processed _per epoch_ (as part the attestation rewards calculation), but proposer rewards are now processed [_per block_](https://github.com/sigp/lighthouse/blob/46be05f7280c3afdfbd614dbc8f6a9b3985ab821/consensus/state_processing/src/per_block_processing/process_operations.rs#L184-L185) since Altair.

It might require [computing the block rewards](https://github.com/sigp/lighthouse/blob/f7bd4bf06ed7e54d9ba55d4c14711cbae367d478/beacon_node/beacon_chain/src/beacon_block_reward.rs#L23) to make this test pass, unless there's a better way to test this.

Reply:
@jimmygchen ah, that makes sense. I will look to add in the computation of the block rewards. I have added in the `process_inactivity_updates` locally to the location you suggested as well.

Reply:
@jimmygchen I tried adding in the computing block rewards logic at the following line:
https://github.com/sigp/lighthouse/blob/d42e9845370b730c17c27bd470e3a641b6764cb9/beacon_node/beacon_chain/tests/rewards.rs#L257-L265

For the above section of code, I am seeing results in this format:
```
Beacon block reward: StandardBlockReward {
    proposer_index: 25,
    total: 0,
    attestations: 0,
    sync_aggregate: 0,
    proposer_slashings: 0,
    attester_slashings: 0,
}
```

I am not seeing a value higher than 0 in the StandardBlockReward response in any of the fields other than the proposer index.

This is a weird issue as I am seeing values in the location:
https://github.com/sigp/lighthouse/blob/46be05f7280c3afdfbd614dbc8f6a9b3985ab821/consensus/state_processing/src/per_block_processing/process_operations.rs#L184-L185

Locally I am printing out the proposer_reward value in the code snippet above, and I am seeing values in that location but not in `compute_beacon_block_reward`

I was wondering if you had any recommendations here?

Reply:
@jimmygchen further investigation has led me to believe that the if statement below is not executing:
https://github.com/sigp/lighthouse/blob/d42e9845370b730c17c27bd470e3a641b6764cb9/beacon_node/beacon_chain/src/beacon_block_reward.rs#L208-L215

I believe the `!validator_participation.has_flag(flag_index)` is the culprit for not allowing block rewards to be calculated. I am not sure what needs to be done to allow for this flag to be correct. Do you have any suggestions here?

Reply:
Hi @zack-scott 

I think the problem here is the `compute_beacon_block_rewards` requires a pre-state (state before applying the block), but the post-state (`current_state`) is being passed to the function. You'd need to use the parent state here instead.

btw I was wrong about phase0 in this comment above
> The reason why the phase0 tests work is because proposer rewards were processed per epoch (as part the attestation rewards calculation), but proposer rewards are now processed [per block](https://github.com/sigp/lighthouse/blob/46be05f7280c3afdfbd614dbc8f6a9b3985ab821/consensus/state_processing/src/per_block_processing/process_operations.rs#L184-L185) since Altair.

The proposer rewards is indeed included as part of the phase 0 attestation rewards calculation function [in the spec](https://github.com/ethereum/consensus-specs/blob/8acb254511d61d3ffcddc262da024646884a8a82/specs/phase0/beacon-chain.md?plain=1#L1517), however the rewards api should exclude the proposer rewards from `inclusion_delay` otherwise we'd be double counting the rewards for users that relies on the block rewards API. I'll fix this in a separate PR.

Reply:
@jimmygchen I am not sure if the way below is the correct way to get the parent state:
```
harness.advance_slot();
let last_block_hash = harness.extend_chain(
    1,
    BlockStrategy::OnCanonicalHead,
    AttestationStrategy::SomeValidators(half_validators.clone()),
).await;
advanced_slot = harness.get_current_slot();

let mut current_state = harness.get_current_state();

let block = harness
    .get_block(SignedBeaconBlockHash::from(last_block_hash))
    .unwrap();

let prev_block = harness
    .get_block(SignedBeaconBlockHash::from(block.parent_root()))
    .unwrap();
let mut previous_state = harness
    .get_hot_state(BeaconStateHash::from(prev_block.state_root()))
    .unwrap();
let beacon_block_reward = harness
    .chain
    .compute_beacon_block_reward(
        prev_block.message(),
        block.parent_root(),
        &mut previous_state,
    )
    .unwrap();
println!("Beacon block reward: {:#?}", beacon_block_reward);
```

The above logic of getting the previous state for computation is still returning 0 balances. I am wondering where the validator participation info is being set? Would this be set via the extend_chain function in the test_utils.rs?

Reply:
For the beacon block reward calculation, you need the pre-state and the **current block**.

Note that `parent_state` is not always the same as pre-state, sorry about using it interchangeably above. They should work most of the time, but they are slightly different at epoch boundaries, because: `pre-state` is the state before applying the block, whereas `parent_state` is the post-state (state after applying a block) of the previous slot.

```mermaid
stateDiagram-v2
    parent_state --> current_pre_state: per slot/epoch processing 
    current_pre_state --> current_post_state: process block 
```

Reply:
@jimmygchen thanks for the explanation. I made use of the `make_block_return_pre_state function on the BeaconChainHarness to compute this pre-state. I was able to get the proposer rewards calculated. I will work on correctly accounting for those values in the test.

Reply:
@jimmygchen I have added in the beacon block rewards to the `expected_balances` calculation, but I am seeing some issues where we get a consistent difference of 89440 over the actual balance. Below values are outputs from current print statements in the test:

```
**************************************************************
Validator: 46 | expected value 31995648660 | slot 56
Validator: 46 | actual value 31995559220 | slot 56
Validator: 46 | difference expected - current: 89440
Validator: 46 | get proposal reward true
**************************************************************
Validator: 47 | expected value 31997546106 | slot 56
Validator: 47 | actual value 31997456666 | slot 56
Validator: 47 | difference expected - current: 89440
Validator: 47 | get proposal reward false
**************************************************************
Validator: 48 | expected value 31994632845 | slot 56
Validator: 48 | actual value 31994632845 | slot 56
Validator: 48 | difference expected - current: 0
Validator: 48 | get proposal reward false
**************************************************************
Validator: 63 | expected value 31995322825 | slot 56
Validator: 63 | actual value 31995322825 | slot 56
Validator: 63 | difference expected - current: 0
Validator: 63 | get proposal reward true
```

We get a mixture of validators that did get the proposal reward and didn't get the proposal reward having this overage. The lines of the code that created this print statement above can be viewed below:
https://github.com/sigp/lighthouse/blob/4cccf2296ed09ac8ea4d9053f39cb1b95d931447/beacon_node/beacon_chain/tests/rewards.rs#L298-L309

I will continue trying to figure out what is causing `exepcted_balances` to be over, but I assume there may be something in the attestation reward calculation that isn't 100% accurate.

Reply:
Hey thanks @zack-scott, I haven't got a chance to look at this today - will continue next week!

Reply:
@jimmygchen the issue was that I also needed to take into account the sync committee penalties. I have added the logic to the test to take these into account, and we have the test passing now. This PR is ready for review now.